lib/galaxy/tools/toolbox/panel.py

```diff
@@ -79,7 +79,10 @@ def to_dict(self, trans, link_details=False):
             link_details=link_details
         )
         for elt in self.elems.values():
-            section_elts.append(elt.to_dict(**kwargs))
+            if elt.__class__.__name__ == 'Tool':
+                section_elts.append(trans.app.toolbox.get_tool_to_dict(trans, elt))
```

Comment:
this fails unit test since the mock trans does not have an app, ideas?

Reply:
Pass a new method parameter along with this call ``get_tool_to_dict=self.get_tool_to_dict``. Is this only used from the toolbox?

Reply:
It may seem like going out of your way but it actually fixes a threading issue where app.toolbox might not be self and I think it has nice cohesion that way as well.

Reply:
Do you mean something like this?
```diff
diff --git a/lib/galaxy/tools/toolbox/base.py b/lib/galaxy/tools/toolbox/base.py
index fe3dc839d7..c38383ce00 100644
--- a/lib/galaxy/tools/toolbox/base.py
+++ b/lib/galaxy/tools/toolbox/base.py
@@ -957,7 +957,7 @@ class AbstractToolBox(Dictifiable, ManagesIntegratedToolPanelMixin):
                 if elt.__class__.__name__ == 'Tool':
                     rval.append(self.get_tool_to_dict(trans, elt))
                 else:
-                    kwargs = dict(trans=trans, link_details=True)
+                    kwargs = dict(trans=trans, link_details=True, get_tool_to_dict=self.get_tool_to_dict)
                     rval.append(elt.to_dict(**kwargs))
         else:
             filter_method = self._build_filter_method(trans)
diff --git a/lib/galaxy/tools/toolbox/panel.py b/lib/galaxy/tools/toolbox/panel.py
index cf547605c9..fc914b70c7 100644
--- a/lib/galaxy/tools/toolbox/panel.py
+++ b/lib/galaxy/tools/toolbox/panel.py
@@ -69,7 +69,7 @@ class ToolSection(Dictifiable, HasPanelItems, object):
         copy.elems = self.elems.copy()
         return copy

-    def to_dict(self, trans, link_details=False):
+    def to_dict(self, trans, link_details=False, get_tool_to_dict=None):
         """ Return a dict that includes section's attributes. """

         section_dict = super(ToolSection, self).to_dict()
@@ -79,8 +79,8 @@ class ToolSection(Dictifiable, HasPanelItems, object):
             link_details=link_details
         )
         for elt in self.elems.values():
-            if elt.__class__.__name__ == 'Tool':
-                section_elts.append(trans.app.toolbox.get_tool_to_dict(trans, elt))
+            if elt.__class__.__name__ == 'Tool' and get_tool_to_dict:
+                section_elts.append(get_tool_to_dict(trans, elt))
             else:
                 section_elts.append(elt.to_dict(**kwargs))
         section_dict['elems'] = section_elts
```